apps/passport-server/src/services/userService.ts

```diff
@@ -287,6 +320,77 @@ export class UserService {
 
     return user;
   }
+
+  /**
+   * Updates the version of the legal terms the user agrees to
+   */
+  public async handleAgreeTerms(
+    serializedPCD: SerializedPCD<SemaphoreSignaturePCD>
+  ): Promise<AgreeTermsResult> {
+    const pcd = await SemaphoreSignaturePCDPackage.deserialize(
+      serializedPCD.pcd
+    );
+    if (await SemaphoreSignaturePCDPackage.verify(pcd)) {
+      const parsedPayload = AgreedTermsSchema.safeParse(
+        JSON.parse(pcd.claim.signedMessage)
+      );
+
+      if (!parsedPayload.success) {
+        return {
+          success: false,
+          error: "Invalid terms specified"
+        };
+      }
+
+      const payload = parsedPayload.data;
+      const user = await fetchUserByCommitment(
+        this.context.dbPool,
+        pcd.claim.identityCommitment
+      );
+      if (!user) {
+        return {
+          success: false,
+          error: "User does not exist"
+        };
+      }
+
+      // If the user hasn't already agreed to have their tickets unredacted,
+      // do it now
+      if (
+        payload.version >= UNREDACT_TICKETS_TERMS_VERSION &&
+        user.terms_agreed < UNREDACT_TICKETS_TERMS_VERSION
+      ) {
+        logger(
+          `[USER_SERVICE] Unredacting tickets for email due to accepting version ${payload.version} of legal terms`,
+          user.email
+        );
+        await agreeTermsAndUnredactTickets(
+          this.context.dbPool,
+          user.email,
+          payload.version
+        );
+      } else {
+        logger(
+          `[USER_SERVICE] Updating user to version ${payload.version} of legal terms`,
+          user.email
+        );
+        await upsertUser(this.context.dbPool, {
+          ...user,
+          terms_agreed: payload.version
+        });
+      }
```

Comment:
the logic of this if/else still slightly doesn't make sense to me.

- it accepts agreeing to terms earlier than have already been agreed to
- it accepts agreeing to negative terms
- i can see that you're validating max terms via the schema, but i think validating inside this function also makes sense
- it upserts the user even if they've already accepted the given terms